File: division-algorithm/division_algorithm.py
from __future__ import annotations

from itertools import zip_longest


def term_to_polynomial(term: tuple[float, int]) -> Polynomial:
    c, p = term
    output = [c]
    output.extend([0] * p)
    return Polynomial(output)


class Polynomial:
    def __init__(self, coefficients: list[float]) -> None:
        while coefficients and coefficients[0] == 0:
            coefficients = coefficients[1:]
        self.coefficients = coefficients
        self.full = list(
            zip(coefficients, range(len(coefficients) - 1, -1, -1), strict=True)
        )
        self.degree = len(coefficients) - 1
        if coefficients:
            self.lt = self.full[0]
        else:
            self.lt = (0, 0)

    def show(self) -> str:
        output_string = ""
        if not self.coefficients:
            return "0"
        for c, p in self.full:
            if not c:
                continue
            if c > 0:
                output_string += f" + "
            else:
                output_string += f" - "
            if abs(c) != 1 or p == 0:
                output_string += f"{abs(c)}"
            if p > 1:
                output_string += f"x^{p}"
            elif p == 1:
                output_string += "x"
        return output_string[3:]

    def negate(self) -> Polynomial:
        negated = [-1 * x for x in self.coefficients]
        return Polynomial(negated)

    def add(self, g: Polynomial) -> Polynomial:
        out: list[float] = []
        for c1, c2 in zip_longest(
            self.coefficients[::-1], g.coefficients[::-1], fillvalue=0
        ):
            out.append(c1 + c2)
        out.reverse()
        return Polynomial(out)

    def mult(self, g: Polynomial) -> Polynomial:
        product = Polynomial([])
        for c1, p1 in self.full:
            for c2, p2 in g.full:
                temp = [c1 * c2]
                temp.extend([0] * (p1 + p2))
                product = product.add(Polynomial(temp))
        return product

    def div(self, g: Polynomial) -> tuple[Polynomial, Polynomial]:
        q = Polynomial([])
        r = self
        g_c, g_p = g.lt
        while (r.coefficients) and (g.degree <= r.degree):
            r_c, r_p = r.lt
            ltr_ltg = term_to_polynomial((r_c / g_c, r_p - g_p))
            q = q.add(ltr_ltg)

            r = r.add(ltr_ltg.mult(g).negate())

        return (q, r)

    def eval(self, x: float) -> float:
        summation = 0
        for c, p in self.full:
            summation += c * (x**p)
        return summation

    def rational_roots(self) -> list[tuple[int, int]]:
        roots = []
        for p in factor(self.coefficients[-1]):
            for q in factor(self.coefficients[0]):
                roots.append((p, q))
        return roots

    def rational_roots_eval(self) -> list[tuple[int, int]]:
        zeros = []
        for p, q in self.rational_roots():
            if self.eval(p / q) == 0:
                zeros.append((p, q))
            elif self.eval(-(p / q)) == 0:
                zeros.append((-p, q))
        return zeros

    def differentiate(self, n: int = 1) -> Polynomial:
        f = Polynomial(self.coefficients)
        for _ in range(n):
            df = Polynomial([])
            for c, p in f.full:
                term = term_to_polynomial((c * p, p - 1))
                df = df.add(term)
            f = Polynomial(df.coefficients)
        return df

    def integral(self) -> Polynomial:
        F = Polynomial([])  # noqa: N806
        for c, p in self.full:
            term = term_to_polynomial((c / (p + 1), p + 1))
            F = F.add(term)  # noqa: N806
        return F

    def gcd(self, g: Polynomial) -> Polynomial:
        h = Polynomial(self.coefficients)
        s = Polynomial(g.coefficients)
        while s.coefficients:
            _, rem = h.div(s)
            h = Polynomial(s.coefficients)
            s = Polynomial(rem.coefficients)
        return h

    def newton_rahpson(self, x0: float, n: int = 100) -> float:
        df = self.differentiate()
        for _ in range(n):
            x0 = x0 - (self.eval(x0) / df.eval(x0))
        return x0

    def bisection(self, min: float, max: float, iter: int = 100) -> float:
        def f(x: float) -> float:
            return self.eval(x)

        for _ in range(iter):
            if f(min) * f(max) > 0:
                msg = "min and max have the same sign"
                raise Exception(msg)
            midpoint = (min + max) / 2
            if f(min) * f(midpoint) > 0:
                min = midpoint
            else:
                max = midpoint
        return midpoint


def factor(n: int) -> list[int]:
    factors = []
    for i in range(1, n + 1):
        if n % i == 0:
            factors.append(i)
    return factors


def main() -> None:
    f = Polynomial([1, 2, 1, 1])
    print(f"f = {f.show()}")
    g = Polynomial([2, 1])
    print(f"g = {g.show()}")
    q, r = f.div(g)
    print(f"f = ({q.show()})({g.show()}) + {r.show()}")
    print("-------------------------")
    f = Polynomial([1, 0, 0, 0, -1])
    g = Polynomial([1, 0, 0, 0, 0, 0, -1])
    h = f.gcd(g)
    print(f"gcd({f.show()}, {g.show()}) = {h.show()}")


if __name__ == "__main__":
    main()

File: division-algorithm/divisionAlgorithm.hs
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE TypeFamilies #-}

module Main where

import Data.Bifunctor (first)
import Data.List (partition, sortBy)
import Data.Ord (comparing)
import Numeric.Natural (Natural)

main :: IO ()
main = do
  putStrLn $ "f = " ++ pprint f
  putStrLn $ "g = " ++ pprint g
  let (q, r) = divAlgorithm f g
  putStrLn $ "f = (" ++ pprint q ++ ")(" ++ pprint g ++ ") + " ++ pprint r
  putStrLn "-------------------------"
  putStrLn $ "gcd(" ++ pprint f1 ++ ", " ++ pprint g1 ++ ") = " ++ pprint (polyGCD f1 g1)

f :: (Coefficient a) => Polynomial a
f = newPoly [1, 2, 1, 1]

g :: (Coefficient a) => Polynomial a
g = newPoly [2, 1]

h :: (Coefficient a) => Polynomial a
h = newPoly [1, 0, -2]

f1 :: (Coefficient a, Fractional a) => Polynomial a
f1 = Polynomial [Term 1 4, Term (-1) 0]

g1 :: (Coefficient a, Fractional a) => Polynomial a
g1 = Polynomial [Term 1 6, Term (-1) 0]

h1 :: (Coefficient a, Fractional a) => Polynomial a
h1 = Polynomial [Term 1 3, Term (-3) 1, Term 2 0]

type Coefficient a = (Num a, Eq a, Ord a)

type Power = Natural

data Term a = Term a Power
  deriving (Eq, Show)

newtype Polynomial a = Polynomial [Term a] deriving (Eq, Show)

class PrettyPrintable a where
  pprint :: a -> String

instance (Show a, Coefficient a) => PrettyPrintable (Term a) where
  pprint (Term 0 _) = "0"
  pprint (Term 1 1) = " + " ++ "x"
  pprint (Term 1 pow)
    | pow == 0 = " + " ++ show 1
    | otherwise = " + " ++ "x^" ++ show pow
  pprint (Term coef 1)
    | coef == 0 = []
    | coef < 0 = " - " ++ show (abs coef) ++ "x"
    | otherwise = " + " ++ show coef ++ "x"
  pprint (Term coef 0)
    | coef < 0 = " - " ++ show (abs coef)
    | coef == 0 = []
    | otherwise = " + " ++ show coef
  pprint (Term coef pow)
    | coef < 0 = " - " ++ show (abs coef) ++ "x^" ++ show pow
    | coef == 0 = []
    | otherwise = " + " ++ show coef ++ "x^" ++ show pow

instance (Show a, Num a, Ord a) => PrettyPrintable (Polynomial a) where
  pprint (Polynomial []) = "0"
  pprint (Polynomial ts) = dropPlus $ go ts
    where
      dropPlus = dropWhile (\c -> c == ' ' || c == '+')
      go [] = []
      go ((Term 0 _) : ts) = go ts
      go (t : ts) = pprint t ++ go ts

class Poly a where
  type Field a
  degree :: a -> Power
  negative :: a -> a
  eval :: a -> Field a -> Field a
  mult :: a -> a -> a
  scale :: a -> Field a -> a
  (<**>) :: a -> a -> a
  (<**>) = mult
  raise :: (Integral b) => a -> b -> a
  raise a 0 = a
  raise a n = a <**> raise a (n - 1)
  ($^) :: (Integral b) => a -> b -> a
  ($^) = raise

infixr 5 $^

instance (Coefficient a) => Poly (Term a) where
  type Field (Term a) = a
  degree (Term _ p) = p
  negative (Term c p) = Term (negate c) p
  eval (Term c p) x = c * (x ^ p)
  mult (Term c1 p1) (Term c2 p2) = Term (c1 * c2) (p1 + p2)
  scale (Term c p) scalar = Term (scalar * c) p

instance Functor Term where
  fmap f (Term c p) = Term (f c) p

instance (Coefficient a) => Poly (Polynomial a) where
  type Field (Polynomial a) = a
  degree (Polynomial []) = 0
  degree ts = let (Polynomial ts') = combine ts in maximum (fmap getP ts')
  negative (Polynomial ts) = Polynomial $ map negative ts
  eval (Polynomial []) _ = 0
  eval (Polynomial (t : ts)) x = eval t x + eval (Polynomial ts) x
  mult (Polynomial t1s) (Polynomial t2s) = combine $ Polynomial [mult t1 t2 | t1 <- t1s, t2 <- t2s]
  scale (Polynomial p) scalar = Polynomial $ map (`scale` scalar) p

instance Functor Polynomial where
  fmap f (Polynomial ts) = Polynomial $ map (fmap f) ts

newPoly :: (Coefficient a) => [a] -> Polynomial a
newPoly ps = Polynomial . reverse $ go (reverse ps) 0
  where
    go [] _ = []
    go (t : ts) n
      | t == 0 = go ts (n + 1)
      | otherwise = Term t n : go ts (n + 1)

lt :: (Coefficient a) => Polynomial a -> Term a
lt (Polynomial []) = Term 0 0
lt (Polynomial (x : _)) = x

divAlgorithm :: forall a. (Coefficient a, Fractional a) => Polynomial a -> Polynomial a -> (Polynomial a, Polynomial a)
divAlgorithm _ (Polynomial []) = error "g must be a nonzero polynomial"
divAlgorithm f g = go (Polynomial []) f
  where
    ltg :: Term a
    ltg = lt g
    go :: Polynomial a -> Polynomial a -> (Polynomial a, Polynomial a)
    go q r@(Polynomial rs)
      | null rs = (q, r)
      | degree g > degree r = (q, r)
      | otherwise = go (q <++> ltrltg) (r <++> negative (ltrltg `mult` g))
      where
        ltrltg :: Polynomial a
        ltrltg = Polynomial [lt r `over` ltg]
        over :: (Coefficient a, Fractional a) => Term a -> Term a -> Term a
        (Term c1 p1) `over` (Term c2 p2) = Term (c1 / c2) (p1 - p2)

(///) :: forall a. (Coefficient a, Fractional a) => Polynomial a -> Polynomial a -> (Polynomial a, Polynomial a)
(///) = divAlgorithm

(<++>) :: (Coefficient a) => Polynomial a -> Polynomial a -> Polynomial a
(Polynomial f) <++> (Polynomial g) = combine . Polynomial $ f ++ g

instance (PrettyPrintable (Polynomial a)) => PrettyPrintable (Polynomial a, Polynomial a) where
  pprint (f, g) = pprint f ++ ", remainder " ++ pprint g

polyGCD :: (Coefficient a, Fractional a) => Polynomial a -> Polynomial a -> Polynomial a
polyGCD = go
  where
    go h (Polynomial []) = h
    go h s = let (_, rem) = divAlgorithm h s in go s rem

multipleGCD :: (Coefficient a, Fractional a) => [Polynomial a] -> Polynomial a
multipleGCD = foldl polyGCD (Polynomial [])

getQ :: (Polynomial a, Polynomial a) -> Polynomial a
getQ = fst

getR :: (Polynomial a, Polynomial a) -> Polynomial a
getR = snd

sortTerms :: Polynomial a -> Polynomial a
sortTerms (Polynomial ts) = Polynomial $ sortBy (flip (comparing getP)) ts

getC :: Term a -> a
getC (Term c _) = c

getP :: Term a -> Power
getP (Term _ p) = p

getCoefficients :: (Coefficient a) => Polynomial a -> [a]
getCoefficients (Polynomial []) = []
getCoefficients ts = go ps maxpower
  where
    (Polynomial ps) = sortTerms $ combine ts
    maxpower = degree ts
    go [] _ = []
    go (x : xs) n
      | getP x == n = getC x : go xs (n - 1)
      | otherwise = 0 : go (x : xs) (n - 1)

combine :: (Num a, Eq a) => Polynomial a -> Polynomial a
combine (Polynomial ts) = sortTerms $ Polynomial (go ts)
  where
    go [] = []
    go (t : ts) =
      let (same, rest) = partition (\x -> getP x == getP t) ts
          coeffSum = sum (map getC (t : same))
       in if coeffSum == 0
            then go rest
            else Term coeffSum (getP t) : go rest

rationalRoots :: Polynomial Integer -> [(Integer, Integer)]
rationalRoots (Polynomial []) = []
rationalRoots (Polynomial ts) =
  [ (p, q)
    | p <- factor $ abs $ getC (last ts),
      q <- factor $ abs $ getC (head ts)
  ]
  where
    factor n = [k | k <- [1 .. n], n `mod` k == 0]

rationalRootsEval :: Polynomial Integer -> [(Integer, Integer)]
rationalRootsEval poly =
  filter (\x -> abs (eval (fmap fromInteger poly) (toDecimal x)) < 0.000001) $ roots ++ fmap (first negate) roots
  where
    roots = rationalRoots poly
    toDecimal :: (Integer, Integer) -> Double
    toDecimal (n, d) = fromInteger n / fromInteger d

class Differentiable a where
  diff :: a -> a
  d :: Int -> a -> a

instance (Coefficient a) => Differentiable (Term a) where
  diff (Term _ 0) = Term 0 0
  diff (Term coef pow) = Term (coef * (fromInteger . toInteger) pow) (pow - 1)
  d n f = iterate diff f !! n

instance (Coefficient a) => Differentiable (Polynomial a) where
  diff (Polynomial ts) = Polynomial $ map diff ts
  d n f = iterate diff f !! n

class (Poly a) => Integrable a where
  integral :: a -> a
  integrate :: a -> Field a -> Field a -> Field a

instance (Coefficient a, Fractional a) => Integrable (Term a) where
  integral (Term coef pow) = Term (coef / fromIntegral (pow + 1)) (pow + 1)
  integrate term lowerBound upperBound =
    let intF = integral term
     in eval intF upperBound - eval intF lowerBound

instance (Coefficient a, Fractional a) => Integrable (Polynomial a) where
  integral (Polynomial ts) = Polynomial $ map integral ts
  integrate poly lowerBound upperBound =
    let intF = integral poly
     in eval intF upperBound - eval intF lowerBound

newtonRahpson :: forall a. (Coefficient a, Fractional a) => Polynomial a -> a -> Integer -> a
newtonRahpson f x0 iter = go x0 1
  where
    df = diff f
    newton x = x - (eval f x / eval df x)
    go x n
      | n > iter = x
      | otherwise = go (newton x) (n + 1)

bisection :: (Coefficient a, Fractional a) => Polynomial a -> a -> a -> Integer -> Maybe a
bisection p min max iter = go p min max 1
  where
    go p min max n
      | n > iter = Just m
      | f min * f max > 0 = Nothing
      | f min * f m < 0 = go p min m (n + 1)
      | otherwise = go p m max (n + 1)
      where
        f = eval p
        m = (min + max) / 2
